File: packages/exchange-adapter/src/exchanges/okx/WsUserDataAdapter.ts
import { WebsocketClient } from 'okx-api'
import { BaseWsUserDataAdapter } from '../../core/BaseWsUserDataAdapter'
import type {
  Exchange,
  TradeType,
  WsEventHandler,
  WsSubscribeOptions,
  WsOrderUpdate,
  WsStrategyOrderUpdate,
  WsPositionUpdate,
  WsBalanceUpdate,
  WsAccountUpdate,
  OrderStatus,
  OrderSide,
  PositionSide,
  OrderType,
  StrategyOrderType,
  StrategyOrderStatus,
} from '../../core/types'
import { okxToUnified } from './utils'
import { createProxyAgent } from '../../core/utils'

// ============================================================================
// OKX WebSocket 响应类型
// ============================================================================

interface OkxWsOrderData {
  instId: string
  instType: string
  ordId: string
  clOrdId?: string
  side: string
  posSide?: string
  ordType: string
  state: string
  px: string
  sz: string
  fillSz: string
  avgPx?: string
  fee?: string
  feeCcy?: string
  reduceOnly?: string
  uTime: string
}

interface OkxWsAlgoOrderData {
  instId: string
  instType: string
  algoId: string
  algoClOrdId?: string
  side: string
  posSide?: string
  ordType: string
  state: string
  triggerPx: string
  ordPx?: string
  sz: string
  triggerTime?: string
  uTime: string
}

interface OkxWsPositionData {
  instId: string
  instType: string
  posSide: string
  pos: string
  avgPx: string
  upl: string
  lever?: string
  mgnMode: string
  liqPx?: string
  uTime: string
}

interface OkxWsBalanceData {
  ccy: string
  cashBal: string
  availBal: string
  frozenBal: string
  upl?: string
  uTime: string
}

interface OkxWsAccountData {
  details: OkxWsBalanceData[]
  uTime: string
}

// ============================================================================
// 初始化参数
// ============================================================================

export interface OkxWsUserDataAdapterInit {
  apiKey: string
  apiSecret: string
  passphrase: string
  /** 是否使用模拟盘 */
  demonet?: boolean
  httpsProxy?: string
  socksProxy?: string
}

// ============================================================================
// OKX WebSocket 用户数据适配器
// ============================================================================

export class OkxWsUserDataAdapter extends BaseWsUserDataAdapter {
  readonly exchange: Exchange = 'okx'

  private wsClient: WebsocketClient
  private connectedTradeTypes: Set<TradeType> = new Set()
  private subscribeOptions: Map<TradeType, WsSubscribeOptions> = new Map()
  /** 是否使用模拟盘 */
  readonly demonet: boolean

  constructor(config: OkxWsUserDataAdapterInit) {
    super()
    this.demonet = config.demonet || false

    this.wsClient = new WebsocketClient({
      accounts: [{
        apiKey: config.apiKey,
        apiSecret: config.apiSecret,
        apiPass: config.passphrase
      }],
      demoTrading: this.demonet,
      wsOptions: {
        agent: createProxyAgent({ socksProxy: config.socksProxy })
      }
    })

    this.setupEventHandlers()
  }

  // ============================================================================
  // 连接状态
  // ============================================================================

  isConnected(tradeType?: TradeType): boolean {
    if (tradeType) {
      return this.connectedTradeTypes.has(tradeType)
    }
    return this.connectedTradeTypes.size > 0
  }

  // ============================================================================
  // 订阅/取消订阅
  // ============================================================================

  async subscribe(options: WsSubscribeOptions, handler: WsEventHandler): Promise<void> {
    const { tradeType, autoReconnect, reconnectInterval, maxReconnectAttempts } = options

    // 更新重连配置
    if (autoReconnect !== undefined) this.reconnectConfig.autoReconnect = autoReconnect
    if (reconnectInterval !== undefined) this.reconnectConfig.reconnectInterval = reconnectInterval
    if (maxReconnectAttempts !== undefined) this.reconnectConfig.maxReconnectAttempts = maxReconnectAttempts

    // 添加全局处理器
    this.addGlobalHandler(handler)

    // 保存订阅选项
    this.subscribeOptions.set(tradeType, options)

    // OKX 使用统一的私有 WebSocket，订阅不同频道
    // 订阅订单频道
    this.wsClient.subscribe({
      channel: 'orders',
      instType: this.getOkxInstType(tradeType)
    })

    // 订阅策略订单频道
    this.wsClient.subscribe({
      channel: 'orders-algo',
      instType: this.getOkxInstType(tradeType)
    })

    // 订阅持仓频道 (合约)
    if (tradeType !== 'spot') {
      this.wsClient.subscribe({
        channel: 'positions',
        instType: this.getOkxInstType(tradeType)
      })
    }

    // 订阅账户频道
    this.wsClient.subscribe({
      channel: 'account'
    })
  }

  async unsubscribe(tradeType?: TradeType): Promise<void> {
    if (tradeType) {
      // 取消特定交易类型的订阅
      this.wsClient.unsubscribe({
        channel: 'orders',
        instType: this.getOkxInstType(tradeType)
      })
      this.wsClient.unsubscribe({
        channel: 'orders-algo',
        instType: this.getOkxInstType(tradeType)
      })
      if (tradeType !== 'spot') {
        this.wsClient.unsubscribe({
          channel: 'positions',
          instType: this.getOkxInstType(tradeType)
        })
      }
      this.subscribeOptions.delete(tradeType)
      this.connectedTradeTypes.delete(tradeType)
    } else {
      // 取消所有订阅
      for (const tt of this.subscribeOptions.keys()) {
        await this.unsubscribe(tt)
      }
    }
  }

  async close(): Promise<void> {
    this.wsClient.closeAll()
    this.connectedTradeTypes.clear()
    this.subscribeOptions.clear()
    this.clearAllListeners()
  }

  // ============================================================================
  // 事件处理
  // ============================================================================

  private setupEventHandlers(): void {
    // 连接状态
    this.wsClient.on('open', (event) => {
      console.log('[OKX WS] Connected:', event.wsKey)
    })

    this.wsClient.on('authenticated', (event) => {
      console.log('[OKX WS] Authenticated:', event.wsKey)
      // 标记所有订阅的交易类型为已连接
      for (const tradeType of this.subscribeOptions.keys()) {
        this.connectedTradeTypes.add(tradeType)
        this.emitConnected(tradeType)
      }
    })

    this.wsClient.on('close', (event) => {
      console.log('[OKX WS] Disconnected:', event.wsKey)
      for (const tradeType of this.connectedTradeTypes) {
        this.emitDisconnected(tradeType, 'Connection closed')
      }
      this.connectedTradeTypes.clear()
    })

    this.wsClient.on('exception', (error: { message?: string }) => {
      console.error('[OKX WS] Error:', error)
      this.emitError('WS_ERROR', error?.message || 'Unknown error')
    })

    // 数据更新
    this.wsClient.on('update', (data: any) => {
      this.handleUpdate(data)
    })
  }

  private handleUpdate(data: any): void {
    const { arg, data: updates } = data
    if (!arg || !updates || !Array.isArray(updates)) return

    const channel = arg.channel
    const instType = arg.instType

    switch (channel) {
      case 'orders':
        updates.forEach((order: OkxWsOrderData) => {
          this.handleOrderUpdate(order, instType)
        })
        break

      case 'orders-algo':
        updates.forEach((algoOrder: OkxWsAlgoOrderData) => {
          this.handleAlgoOrderUpdate(algoOrder, instType)
        })
        break

      case 'positions':
        updates.forEach((position: OkxWsPositionData) => {
          this.handlePositionUpdate(position, instType)
        })
        break

      case 'account':
        if (updates[0]) {
          this.handleAccountUpdate(updates[0] as OkxWsAccountData, instType)
        }
        break
    }
  }

  private handleOrderUpdate(order: OkxWsOrderData, instType: string): void {
    const tradeType = this.instTypeToTradeType(instType)
    const { symbol } = okxToUnified(order.instId)

    const update: WsOrderUpdate = {
      eventType: 'order',
      symbol,
      tradeType,
      orderId: order.ordId,
      clientOrderId: order.clOrdId,
      side: order.side.toLowerCase() as OrderSide,
      positionSide: order.posSide?.toLowerCase() as PositionSide,
      orderType: this.mapOrderType(order.ordType),
      status: this.mapOrderStatus(order.state),
      price: order.px,
      quantity: order.sz,
      filledQuantity: order.fillSz,
      avgPrice: order.avgPx,
      fee: order.fee,
      feeAsset: order.feeCcy,
      reduceOnly: order.reduceOnly === 'true',
      updateTime: parseInt(order.uTime),
      raw: order
    }

    this.emitOrderUpdate(update)
  }

  private handleAlgoOrderUpdate(algoOrder: OkxWsAlgoOrderData, instType: string): void {
    const tradeType = this.instTypeToTradeType(instType)
    const { symbol } = okxToUnified(algoOrder.instId)

    const update: WsStrategyOrderUpdate = {
      eventType: 'strategyOrder',
      symbol,
      tradeType,
      algoId: algoOrder.algoId,
      clientAlgoId: algoOrder.algoClOrdId,
      side: algoOrder.side.toLowerCase() as OrderSide,
      positionSide: algoOrder.posSide?.toLowerCase() as PositionSide,
      strategyType: this.mapStrategyOrderType(algoOrder.ordType),
      status: this.mapStrategyOrderStatus(algoOrder.state),
      triggerPrice: algoOrder.triggerPx,
      orderPrice: algoOrder.ordPx,
      quantity: algoOrder.sz,
      triggerTime: algoOrder.triggerTime ? parseInt(algoOrder.triggerTime) : undefined,
      updateTime: parseInt(algoOrder.uTime),
      raw: algoOrder
    }

    this.emitStrategyOrderUpdate(update)
  }

  private handlePositionUpdate(position: OkxWsPositionData, instType: string): void {
    const tradeType = this.instTypeToTradeType(instType)
    const { symbol } = okxToUnified(position.instId)

    const update: WsPositionUpdate = {
      eventType: 'position',
      symbol,
      tradeType,
      positionSide: position.posSide.toLowerCase() as PositionSide,
      quantity: position.pos,
      entryPrice: position.avgPx,
      unrealizedPnl: position.upl,
      leverage: position.lever,
      marginType: position.mgnMode === 'cross' ? 'cross' : 'isolated',
      liquidationPrice: position.liqPx,
      updateTime: parseInt(position.uTime),
      raw: position
    }

    this.emitPositionUpdate(update)
  }

  private handleAccountUpdate(account: OkxWsAccountData, instType: string): void {
    const tradeType = this.instTypeToTradeType(instType || 'SWAP')
    const updateTime = parseInt(account.uTime)

    // 发送单独的余额更新
    account.details.forEach((balance) => {
      const balanceUpdate: WsBalanceUpdate = {
        eventType: 'balance',
        asset: balance.ccy,
        tradeType,
        available: balance.availBal,
        total: balance.cashBal,
        frozen: balance.frozenBal,
        unrealizedPnl: balance.upl,
        updateTime: parseInt(balance.uTime),
        raw: balance
      }
      this.emitBalanceUpdate(balanceUpdate)
    })

    // 发送综合账户更新
    const accountUpdate: WsAccountUpdate = {
      eventType: 'account',
      tradeType,
      balances: account.details.map((balance) => ({
        asset: balance.ccy,
        tradeType,
        available: balance.availBal,
        total: balance.cashBal,
        frozen: balance.frozenBal,
        unrealizedPnl: balance.upl,
        updateTime: parseInt(balance.uTime),
        raw: balance
      })),
      positions: [], // 持仓通过单独的 positions 频道推送
      updateTime,
      raw: account
    }

    this.emitAccountUpdate(accountUpdate)
  }

  // ============================================================================
  // 辅助方法
  // ============================================================================

  private getOkxInstType(tradeType: TradeType): 'SPOT' | 'SWAP' | 'FUTURES' {
    switch (tradeType) {
      case 'spot':
        return 'SPOT'
      case 'futures':
        return 'SWAP'
      case 'delivery':
        return 'FUTURES'
      default:
        return 'SWAP'
    }
  }

  private instTypeToTradeType(instType: string): TradeType {
    switch (instType?.toUpperCase()) {
      case 'SPOT':
        return 'spot'
      case 'SWAP':
        return 'futures'
      case 'FUTURES':
        return 'delivery'
      default:
        return 'futures'
    }
  }

  private mapOrderType(ordType: string): OrderType {
    switch (ordType?.toLowerCase()) {
      case 'market':
        return 'market'
      case 'limit':
        return 'limit'
      case 'post_only':
        return 'maker-only'
      case 'fok':
      case 'ioc':
        return 'limit' // FOK/IOC are execution types, base order is limit
      default:
        return 'limit'
    }
  }

  private mapOrderStatus(state: string): OrderStatus {
    switch (state?.toLowerCase()) {
      case 'live':
        return 'open'
      case 'partially_filled':
        return 'partial'
      case 'filled':
        return 'filled'
      case 'canceled':
      case 'cancelled':
        return 'canceled'
      default:
        return 'open'
    }
  }

  private mapStrategyOrderType(ordType: string): StrategyOrderType {
    switch (ordType?.toLowerCase()) {
      case 'conditional':
        return 'stop-loss'
      case 'oco':
        return 'take-profit'
      case 'trigger':
        return 'trigger'
      case 'move_order_stop':
        return 'trailing-stop'
      default:
        return 'trigger'
    }
  }

  private mapStrategyOrderStatus(state: string): StrategyOrderStatus {
    switch (state?.toLowerCase()) {
      case 'live':
        return 'live'
      case 'effective':
        return 'effective'
      case 'canceled':
      case 'cancelled':
        return 'canceled'
      case 'order_failed':
        return 'failed'
      case 'partially_effective':
        return 'partially_effective'
      default:
        return 'live'
    }
  }
}

File: pkg/exchange-adapter/exchanges/okx/ws_user_data_adapter.go
package okx

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/pkg/exchange-adapter/core"
	okxapi "github.com/pkg/okx-api"
	okxtypes "github.com/pkg/okx-api/types"
	okxws "github.com/pkg/okx-api/types/websockets"
)

type WsUserDataAdapterOptions struct {
	APIKey    string
	APISecret string
	APIPass   string
	Options   *core.AdapterOptions
}

type WsUserDataAdapter struct {
	core.BaseWsUserDataAdapter

	ws *okxapi.WebsocketClient

	mu            sync.Mutex
	subscribed    map[core.TradeType][]core.ListenerID
	connected     map[core.TradeType]bool
	accountSubbed bool
}

func NewWsUserDataAdapter(opts WsUserDataAdapterOptions) *WsUserDataAdapter {
	demonet := true
	if opts.Options != nil && opts.Options.Demonet != nil {
		demonet = *opts.Options.Demonet
	}

	proxyURL := ""
	socksProxyURL := ""
	if opts.Options != nil {
		proxyURL = opts.Options.HTTPSProxy
		socksProxyURL = opts.Options.SOCKSProxy
	}

	ws := okxapi.NewWebsocketClient(okxws.WSClientConfig{
		Accounts: []okxtypes.APICredentials{{
			APIKey:    opts.APIKey,
			APISecret: opts.APISecret,
			APIPass:   opts.APIPass,
		}},
		Market:           okxtypes.APIMarketGLOBAL,
		DemoTrading:      demonet,
		ReconnectTimeout: 5 * time.Second,
		Proxy:            proxyURL,
		SocksProxy:       socksProxyURL,
	})

	a := &WsUserDataAdapter{
		BaseWsUserDataAdapter: core.NewBaseWsUserDataAdapter(core.ExchangeOKX),
		ws:                    ws,
		subscribed:            make(map[core.TradeType][]core.ListenerID),
		connected:             make(map[core.TradeType]bool),
	}

	a.setupEventHandlers()
	return a
}

func (a *WsUserDataAdapter) IsConnected(tradeType *core.TradeType) bool {
	a.mu.Lock()
	defer a.mu.Unlock()
	if tradeType == nil {
		return len(a.connected) > 0
	}
	return a.connected[*tradeType]
}

func (a *WsUserDataAdapter) Subscribe(ctx context.Context, options core.WsSubscribeOptions, handler core.WsEventHandler) error {
	a.applySubscribeOptions(options)

	a.mu.Lock()
	tt := options.TradeType
	a.subscribed[tt] = append(a.subscribed[tt], a.OnAny(handler))

	needAccount := !a.accountSubbed
	a.accountSubbed = true
	a.mu.Unlock()

	args := make([]map[string]any, 0, 4)
	args = append(args, map[string]any{
		"channel":  "orders",
		"instType": tradeTypeToInstType(tt),
	})
	args = append(args, map[string]any{
		"channel":  "orders-algo",
		"instType": tradeTypeToInstType(tt),
	})
	if tt != core.TradeTypeSpot {
		args = append(args, map[string]any{
			"channel":  "positions",
			"instType": tradeTypeToInstType(tt),
		})
	}
	if needAccount {
		args = append(args, map[string]any{"channel": "account"})
	}

	return a.ws.Subscribe(ctx, args, nil)
}

func (a *WsUserDataAdapter) Unsubscribe(ctx context.Context, tradeType *core.TradeType) error {
	if tradeType == nil {
		a.mu.Lock()
		keys := make([]core.TradeType, 0, len(a.subscribed))
		for tt := range a.subscribed {
			keys = append(keys, tt)
		}
		a.mu.Unlock()

		for _, tt := range keys {
			ttCopy := tt
			_ = a.Unsubscribe(ctx, &ttCopy)
		}
		return nil
	}

	tt := *tradeType

	a.mu.Lock()
	ids := a.subscribed[tt]
	delete(a.subscribed, tt)
	delete(a.connected, tt)
	noSubs := len(a.subscribed) == 0
	if noSubs {
		a.accountSubbed = false
	}
	a.mu.Unlock()

	for _, id := range ids {
		a.OffAny(id)
	}

	args := make([]map[string]any, 0, 3)
	args = append(args, map[string]any{
		"channel":  "orders",
		"instType": tradeTypeToInstType(tt),
	})
	args = append(args, map[string]any{
		"channel":  "orders-algo",
		"instType": tradeTypeToInstType(tt),
	})
	if tt != core.TradeTypeSpot {
		args = append(args, map[string]any{
			"channel":  "positions",
			"instType": tradeTypeToInstType(tt),
		})
	}

	_ = a.ws.Unsubscribe(ctx, args, nil)

	if noSubs {
		_ = a.ws.Unsubscribe(ctx, []map[string]any{{"channel": "account"}}, nil)
		a.ws.CloseAll()
	}

	return nil
}

func (a *WsUserDataAdapter) Close(ctx context.Context) error {
	_ = a.Unsubscribe(ctx, nil)
	a.ws.CloseAll()
	return nil
}

// ============================================================================//
// OKX websocket events
// ============================================================================//

func (a *WsUserDataAdapter) setupEventHandlers() {
	a.ws.On("authenticated", func(_ any) {
		a.mu.Lock()
		tradeTypes := make([]core.TradeType, 0, len(a.subscribed))
		for tt := range a.subscribed {
			tradeTypes = append(tradeTypes, tt)
		}
		a.mu.Unlock()

		for _, tt := range tradeTypes {
			a.mu.Lock()
			if a.connected[tt] {
				a.mu.Unlock()
				continue
			}
			a.connected[tt] = true
			a.mu.Unlock()

			ttCopy := tt
			a.EmitConnected(&ttCopy)
		}
	})

	a.ws.On("close", func(_ any) {
		a.mu.Lock()
		tradeTypes := make([]core.TradeType, 0, len(a.connected))
		for tt := range a.connected {
			tradeTypes = append(tradeTypes, tt)
		}
		a.connected = make(map[core.TradeType]bool)
		a.mu.Unlock()

		for _, tt := range tradeTypes {
			ttCopy := tt
			a.EmitDisconnected(&ttCopy, "connection closed")
		}
	})

	a.ws.OnError(func(err error) {
		if err == nil {
			return
		}
		a.EmitError(core.ErrorWsConnection, err.Error(), nil, err)
	})

	a.ws.On("exception", func(event any) {
		msg := ""
		switch v := event.(type) {
		case okxapi.WsEvent:
			msg = v.Message.Msg
			if msg == "" {
				msg = string(v.Raw)
			}
		case okxapi.WsRawMessage:
			msg = string(v.Raw)
		default:
			msg = fmt.Sprintf("%v", v)
		}
		a.EmitError(core.ErrorWsSubscription, msg, nil, event)
	})

	a.ws.On("update", func(event any) {
		upd, ok := event.(okxapi.WsDataEvent)
		if !ok {
			return
		}
		a.handleUpdate(upd)
	})
}

type okxWsArg struct {
	Channel  string `json:"channel"`
	InstType string `json:"instType,omitempty"`
}

type okxWsOrder struct {
	InstID     string `json:"instId"`
	OrdID      string `json:"ordId"`
	ClOrdID    string `json:"clOrdId"`
	Side       string `json:"side"`
	PosSide    string `json:"posSide"`
	OrdType    string `json:"ordType"`
	State      string `json:"state"`
	Px         string `json:"px"`
	Sz         string `json:"sz"`
	FillSz     string `json:"fillSz"`
	AvgPx      string `json:"avgPx"`
	Fee        string `json:"fee"`
	FeeCcy     string `json:"feeCcy"`
	ReduceOnly string `json:"reduceOnly"`
	UTime      string `json:"uTime"`
}

type okxWsAlgoOrder struct {
	InstID      string `json:"instId"`
	AlgoID      string `json:"algoId"`
	AlgoClOrdID string `json:"algoClOrdId"`
	Side        string `json:"side"`
	PosSide     string `json:"posSide"`
	OrdType     string `json:"ordType"`
	State       string `json:"state"`
	TriggerPx   string `json:"triggerPx"`
	OrdPx       string `json:"ordPx"`
	Sz          string `json:"sz"`
	TriggerTime string `json:"triggerTime"`
	UTime       string `json:"uTime"`
}

type okxWsPosition struct {
	InstID  string `json:"instId"`
	PosSide string `json:"posSide"`
	Pos     string `json:"pos"`
	AvgPx   string `json:"avgPx"`
	Upl     string `json:"upl"`
	Lever   string `json:"lever"`
	MgnMode string `json:"mgnMode"`
	LiqPx   string `json:"liqPx"`
	UTime   string `json:"uTime"`
}

type okxWsBalance struct {
	Ccy       string `json:"ccy"`
	CashBal   string `json:"cashBal"`
	AvailBal  string `json:"availBal"`
	FrozenBal string `json:"frozenBal"`
	Upl       string `json:"upl"`
	UTime     string `json:"uTime"`
}

type okxWsAccount struct {
	Details []okxWsBalance `json:"details"`
	UTime   string         `json:"uTime"`
}

func (a *WsUserDataAdapter) handleUpdate(upd okxapi.WsDataEvent) {
	var arg okxWsArg
	if err := json.Unmarshal(upd.Message.Arg, &arg); err != nil {
		return
	}

	tradeType := core.TradeTypeFutures
	if arg.InstType != "" {
		tradeType = instTypeToTradeType(arg.InstType)
	}

	switch arg.Channel {
	case "orders":
		var data []okxWsOrder
		if err := json.Unmarshal(upd.Message.Data, &data); err != nil {
			return
		}
		for _, o := range data {
			update := core.WsOrderUpdate{
				Symbol:         rawToUnifiedSymbol(o.InstID),
				TradeType:      tradeType,
				OrderID:        o.OrdID,
				ClientOrderID:  o.ClOrdID,
				Side:           toCoreSide(o.Side),
				PositionSide:   toCorePositionSide(o.PosSide),
				OrderType:      toCoreOrderType(o.OrdType),
				Status:         toCoreOrderStatus(o.State),
				Price:          o.Px,
				Quantity:       o.Sz,
				FilledQuantity: o.FillSz,
				AvgPrice:       o.AvgPx,
				Fee:            o.Fee,
				FeeAsset:       o.FeeCcy,
				ReduceOnly:     strings.EqualFold(strings.TrimSpace(o.ReduceOnly), "true"),
				UpdateTime:     msStringToInt64(o.UTime),
				Raw:            json.RawMessage(upd.Raw),
			}
			a.Emit(update)
		}

	case "orders-algo":
		var data []okxWsAlgoOrder
		if err := json.Unmarshal(upd.Message.Data, &data); err != nil {
			return
		}
		for _, o := range data {
			var triggerTime *int64
			if v := msStringToInt64(o.TriggerTime); v > 0 {
				triggerTime = &v
			}
			update := core.WsStrategyOrderUpdate{
				Symbol:       rawToUnifiedSymbol(o.InstID),
				TradeType:    tradeType,
				AlgoID:       o.AlgoID,
				ClientAlgoID: o.AlgoClOrdID,
				Side:         toCoreSide(o.Side),
				PositionSide: toCorePositionSide(o.PosSide),
				StrategyType: toCoreStrategyOrderType(o.OrdType, false),
				Status:       toCoreStrategyOrderStatus(o.State),
				TriggerPrice: o.TriggerPx,
				OrderPrice:   o.OrdPx,
				Quantity:     o.Sz,
				TriggerTime:  triggerTime,
				UpdateTime:   msStringToInt64(o.UTime),
				Raw:          json.RawMessage(upd.Raw),
			}
			a.Emit(update)
		}

	case "positions":
		var data []okxWsPosition
		if err := json.Unmarshal(upd.Message.Data, &data); err != nil {
			return
		}
		for _, p := range data {
			update := core.WsPositionUpdate{
				Symbol:           rawToUnifiedSymbol(p.InstID),
				TradeType:        tradeType,
				PositionSide:     core.PositionSide(strings.ToLower(p.PosSide)),
				Quantity:         p.Pos,
				EntryPrice:       p.AvgPx,
				UnrealizedPnl:    p.Upl,
				Leverage:         p.Lever,
				MarginType:       core.MarginMode(strings.ToLower(p.MgnMode)),
				LiquidationPrice: p.LiqPx,
				UpdateTime:       msStringToInt64(p.UTime),
				Raw:              json.RawMessage(upd.Raw),
			}
			a.Emit(update)
		}

	case "account":
		var data []okxWsAccount
		if err := json.Unmarshal(upd.Message.Data, &data); err != nil {
			return
		}
		for _, acct := range data {
			updateTime := msStringToInt64(acct.UTime)
			balances := make([]core.WsBalanceUpdate, 0, len(acct.Details))
			for _, b := range acct.Details {
				bu := core.WsBalanceUpdate{
					Asset:         b.Ccy,
					TradeType:     tradeType,
					Available:     b.AvailBal,
					Total:         b.CashBal,
					Frozen:        b.FrozenBal,
					UnrealizedPnl: b.Upl,
					UpdateTime:    msStringToInt64(b.UTime),
					Raw:           json.RawMessage(upd.Raw),
				}
				balances = append(balances, bu)
				a.Emit(bu)
			}

			a.Emit(core.WsAccountUpdate{
				TradeType:  tradeType,
				Balances:   balances,
				Positions:  nil,
				UpdateTime: updateTime,
				Raw:        json.RawMessage(upd.Raw),
			})
		}
	}
}

func (a *WsUserDataAdapter) applySubscribeOptions(options core.WsSubscribeOptions) {
	if options.AutoReconnect != nil {
		a.AutoReconnect = *options.AutoReconnect
	}
	if options.ReconnectInterval != nil {
		a.ReconnectInterval = *options.ReconnectInterval
	}
	if options.MaxReconnectAttempts != nil {
		a.MaxReconnectAttempts = *options.MaxReconnectAttempts
	}
}
